src/cpp/session/modules/environment/SessionEnvironment.cpp

```diff
@@ -112,6 +112,57 @@ bool handleRBrowseEnv(const core::FilePath& filePath)
    }
 }
 
+bool hasExternalPtr(SEXP env,      // environment to search for external pointers
+                    bool nullPtr,  // whether to look for NULL pointers 
+                    int level = 5) // maximum recursion depth (envs can have self-ref loops)
+{
+   // list the contents of this environment
+   std::vector<r::sexp::Variable> vars;
+   r::sexp::Protect rProtect;
+   r::sexp::listEnvironment(env, 
```

Comment:
This will blow up if the object happens to be an S4 object subclassing `"environment"` -- you might want to validate with e.g. `sexp::isPrimitiveEnvironment()`; our implementation of `listEnvironment()` should probably become S4-object aware as well (but unfortunately interacting with S4 objects at the C / C++ level is tough).